### PR TITLE
Fix crash in language service when using on inherited properties of self-extend

### LINKED ISSUE #4581: Semantic operations on property in self-extending interface crashes LS
```TypeScript
interface C extends C {
    /**/prop: number;
}
```

Try renaming and Visual Studio completely crashes.

Try getting highlights/references and you'll get a language service crash:

```
Description:

getDocumentHighlights('C:\Users\drosen\AppData\Local\Temp\~vs4D49.ts', 36)

StackTrace:

Error: Out of stack space
   at getTypeOfNode (:25551:13)
   at getPropertySymbolFromTypeReference (:45793:25)
   at forEach (:860:17)
   at Anonymous function (:45786:29)
   at forEach (:860:17)
   at getPropertySymbolsFromBaseTypes (:45780:21)
   at getPropertySymbolFromTypeReference (:45800:29)
   at forEach (:860:17)
   at Anonymous function (:45786:29)
   at forEach (:860:17)
   at getPropertySymbolsFromBaseTypes (:45780:21)
   at getPropertySymbolFromTypeReference (:45800:29)
   at forEach (:860:17)
   at Anonymous function (:45786:29)
   at forEach (:860:17)
   at getPropertySymbolsFromBaseTypes (:45780:21)
   at getPropertySymbolFromTypeReference (:45800:29)
   at forEach (:860:17)
   at Anonymous function (:45786:29)
   at forEach (:860:17)
   at getPropertySymbolsFromBaseTypes (:45780:21)
   at getPropertySymbolFromTypeReference (:45800:29)
   at forEac......

```

### PR DESCRIPTION
**Pull request by yuit**
Fix #4581

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
This is not a sufficient fix. I can break this by creating an arbitrarily long loop of heritage clauses.

Here's an example:

```ts
interface C extends D {
    /*1*/propC: number;
}

interface D extends C {
    /*2*/propD: string;
}
```

Both occurrence locations cause a crash.

File: src/services/services.ts [hunk 1]
```diff
@@ -6014,7 +6037,7 @@ namespace ts {
                             }
 
                             // Visit the typeReference as well to see if it directly or indirectly use that property
-                            getPropertySymbolsFromBaseTypes(type.symbol, propertyName, result);
+                            getPropertySymbolsFromBaseTypes(type.symbol, propertyName, result, symbol);
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: I think you'll just need to pass along a cache for this; also, can you make the `result` parameter come last?

**Comment by yuit** [comment]
@DanielRosenwasser  any more comment?

File: src/services/services.ts [hunk 2]
```diff
@@ -5983,14 +5983,38 @@ namespace ts {
 
                     // Add symbol of properties/methods of the same name in base classes and implemented interfaces definitions
                     if (rootSymbol.parent && rootSymbol.parent.flags & (SymbolFlags.Class | SymbolFlags.Interface)) {
-                        getPropertySymbolsFromBaseTypes(rootSymbol.parent, rootSymbol.getName(), result);
+                        getPropertySymbolsFromBaseTypes(rootSymbol.parent, rootSymbol.getName(), result, {});
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Add a `/*previousIterationSymbolsCache*/`

File: src/services/services.ts [hunk 3]
```diff
@@ -5983,14 +5983,38 @@ namespace ts {
...
+             * Find symbol of the given property-name and add the symbol to the given result array
+             * @param symbol a symbol to start searching for the given propertyName
+             * @param propertyName a name of property to serach for
+             * @param result an array of symbol of found property symbols
+             * @param previousIterationSymbolsCache a cache of symbol from previous iterations of calling this function to prevent infinite revisitng of the same symbol.
+             *                                The value of previousIterationSymbol is undefined when the function is first called.
+             */
+            function getPropertySymbolsFromBaseTypes(symbol: Symbol, propertyName: string, result: Symbol[],
+                previousIterationSymbolsCache: SymbolTable): void {
+                // If the current symbol is the smae as the previous-iteration symbol, we can just return as the symbol has already been visited
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: `same

File: src/services/services.ts [hunk 4]
```diff
@@ -5983,14 +5983,38 @@ namespace ts {
...
+                // This is particularly important for the following cases, so that we do not inifinitely visit the same symbol.
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: infinitely

File: src/services/services.ts (hunk 3 above)
Comment (DanielRosenwasser) [inline src/services/services.ts]: same

File: src/services/services.ts [hunk 5]
```diff
@@ -6055,7 +6080,7 @@ namespace ts {
                     // see if any is in the list
                     if (rootSymbol.parent && rootSymbol.parent.flags & (SymbolFlags.Class | SymbolFlags.Interface)) {
                         const result: Symbol[] = [];
-                        getPropertySymbolsFromBaseTypes(rootSymbol.parent, rootSymbol.getName(), result);
+                        getPropertySymbolsFromBaseTypes(rootSymbol.parent, rootSymbol.getName(), result, {});
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: /*previousIterationSymbolsCache*/

File: src/services/services.ts [hunk 6]
```diff
@@ -5983,14 +5983,38 @@ namespace ts {
...
+                // This is particularly important for the following cases, so that we do not inifinitely visit the same symbol.
+                // For example:
+                //      interface C extends C {
+                //          /*findRef*/propName: string;
+                //      }
+                // The first time getPropertySymbolsFromBaseTypes is called when finding-all-references at propName,
+                // the symbol argument will be the symbol of an interface "C" and previousIterationSymbol is undefined,
+                // the function will add any found symbol of the property-name, then its sub-routine will call
+                // getPropertySymbolsFromBaseTypes again to walk up any base types to prevent revisiting already
+                // visited symbol, interface "C", the sub- routine will pass the current symbol as previousIterationSymbol.
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: remove space after hyphen

File: src/services/services.ts [hunk 7]
```diff
@@ -5983,14 +5983,38 @@ namespace ts {
...
+                if (previousIterationSymbolsCache && previousIterationSymbolsCache[symbol.name] === symbol) {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Is `previousIterationSymbolsCache` ever not passed in?

Comment (DanielRosenwasser) [inline src/services/services.ts]: Also I think the more appropriate check would be `hasProperty(previousIterationSymbolsCache, symbol.name)`

**Comment by DanielRosenwasser** [comment]
Apart from the comments: :cookie: :cake: :candy: :100:

**Pull request merged**